pkg/expr/convert_to_long.go

```diff
@@ -0,0 +1,312 @@
+	}
+	inputFrame := frames[0]
+
+	// The Frame should have no more than one row
+	if inputFrame.Rows() > 1 {
+		return nil, fmt.Errorf("expected no more than one row in the frame, but got %d", inputFrame.Rows())
+	}
+
+	// Gather:
+	// - unique numeric Field Names, and
+	// - unique Label Keys (from Numeric Fields only)
+	// each one maps to a field in the output long Frame.
+	uniqueNames := make([]string, 0)
+	uniqueKeys := make([]string, 0)
+
+	uniqueNamesMap := make(map[string]data.FieldType)
+	uniqueKeysMap := make(map[string]struct{})
+
+	prints := make(map[string]int)
+
+	registerPrint := func(labels data.Labels) {
+		fp := labels.Fingerprint().String()
+		if _, ok := prints[fp]; !ok {
+			prints[fp] = len(prints)
+		}
+	}
+
+	for _, field := range inputFrame.Fields {
+		if field.Type().Numeric() {
+			if _, ok := uniqueNamesMap[field.Name]; !ok {
+				uniqueNames = append(uniqueNames, field.Name)
+				uniqueNamesMap[field.Name] = field.Type()
+
+			}
+
+			if field.Labels != nil {
+				registerPrint(field.Labels)
+				for key := range field.Labels {
+					if _, ok := uniqueKeysMap[key]; !ok {
+						uniqueKeys = append(uniqueKeys, key)
+					}
+					uniqueKeysMap[key] = struct{}{}
+				}
+			}
+		}
+	}
+
+	// Create new fields for output Long frame
+	fields := make([]*data.Field, 0, len(uniqueNames)+len(uniqueKeys))
+
+	// Create the Numeric Fields, tracking the index of each field by name
+	// Note: May want to use FloatAt and and prepopulate with NaN so missing
+	// combinations of value can be NA instead of the zero value of 0.
+	var nameIndexMap = make(map[string]int, len(uniqueNames))
+	for i, name := range uniqueNames {
+		field := data.NewFieldFromFieldType(uniqueNamesMap[name], len(prints))
+		field.Name = name
+		fields = append(fields, field)
+		nameIndexMap[name] = i
+	}
+
+	// Create the String fields, tracking the index of each field by key
+	var keyIndexMap = make(map[string]int, len(uniqueKeys))
+	for i, k := range uniqueKeys {
+		fields = append(fields, data.NewField(k, nil, make([]string, len(prints))))
+		keyIndexMap[k] = len(nameIndexMap) + i
+	}
+
+	longFrame := data.NewFrame("", fields...)
+
+	if inputFrame.Rows() == 0 {
+		return data.Frames{longFrame}, nil
+	}
+
+	// Add Rows to the fields
+	for _, field := range inputFrame.Fields {
+		if !field.Type().Numeric() {
+			continue
+		}
+		fieldIdx := prints[field.Labels.Fingerprint().String()]
+		longFrame.Fields[nameIndexMap[field.Name]].Set(fieldIdx, field.CopyAt(0))
+		for key, value := range field.Labels {
+			longFrame.Fields[keyIndexMap[key]].Set(fieldIdx, value)
+		}
+	}
+
+	return data.Frames{longFrame}, nil
+}
+
+func convertTimeSeriesMultiToTimeSeriesLong(frames data.Frames) (data.Frames, error) {
+	// Collect all time values and ensure no duplicates
+	timeSet := make(map[time.Time]struct{})
+	labelKeys := make(map[string]struct{})     // Collect all unique label keys
```

Comment:
I'm wondering whether we could make our naming a bit more consistent.
- Elsewhere in this PR we have `uniqueKeysMap := make(map[string]struct{})`
- In `time_series.go` we have `uniqueLabelKeys := make(map[string]struct{})         // unique Label keys, used to build schema`

Not a big deal, but I expect it would be clearer if we align these a bit more.